src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2420,15 +2547,28 @@ private static TypeSymbolWithAnnotations InferResultNullability(BoundExpression
                     }
                     break;
 
+                case ConversionKind.NoConversion:
                 case ConversionKind.Identity:
                 case ConversionKind.DefaultOrNullLiteral:
                 case ConversionKind.ImplicitReference:
                 case ConversionKind.ExplicitReference:
-                    // Inherit state from the operand
-                    // PROTOTYPE(NullableReferenceTypes): Should an explicit cast cast away
-                    // outermost nullability? For instance, is `s` a `string!` or `string?`?
-                    // object? obj = ...; var s = (string)obj;
-                    isNullableIfReferenceType = (operandType is null) ? operand.IsLiteralNullOrDefault() : operandType.IsNullable;
+                    if (operandType is null && operandOpt.IsLiteralNullOrDefault())
+                    {
+                        isNullableIfReferenceType = true;
+                    }
+                    else
+                    {
+                        // Inherit state from the operand
+                        // PROTOTYPE(NullableReferenceTypes): Should an explicit cast cast away
+                        // outermost nullability? For instance, is `s` a `string!` or `string?`?
+                        // object? obj = ...; var s = (string)obj;
+                        if (checkConversion)
+                        {
+                            conversion = GenerateConversion(_conversions, operandOpt, operandType?.TypeSymbol, targetType, fromExplicitCast);
```

Comment:
>GenerateConversion [](start = 41, length = 18)

Can we assert that the re-calculated conversion is "similar" to the conversion we were given?

Reply:
Added `// PROTOTYPE` comment.

---
In reply to: [189390941](https://github.com/dotnet/roslyn/pull/26356#discussion_r189390941) [](ancestors = 189390941)